README.md
# Flask App
Recognize Anything model(RAM) is an image tagging model, which can recognize any common category with high accuracy.

[Grounded Segment Anything Model(GSAM)](https://github.com/chenxwh/Grounded-Segment-Anything) combines [Grounding DINO](https://github.com/IDEA-Research/GroundingDINO) with [Segment Anything](https://github.com/facebookresearch/segment-anything)

[Scene Explain](https://scenex.jina.ai/) explains the image. Different models with different options are available.

# Environment setup, to run the app locally
- ignore this if you're using Dockerfile
- Create a .env file similar to .env.example
- Get replicate API key from [here](https://replicate.com/)
- Get scene explain API key from [here](https://scenex.jina.ai/)
- Fill in some secret key

# Run app locally
```
python run.py
```

# Run app using Docker

In the root dir(btb_ai/),  run
```
docker build -t btb_ai:version .
docker run -p 8000:5000 btb_ai:version

```
For current version, check CHANGELOG.md

# Tests

run this at the root dir
```
pytest
```

# Packaging and Distribution

pending...


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This format adheres to [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines.


## [0.1.0] - 21.07.2023

### Added
- [Feature] Added Scene Explain API

## [0.0.1] - 14.07.2023

### Added
- [Feature] Added Replicate API


tests/test_ai_model.py
import json
import pytest
import requests_mock
import sys

from pathlib import Path

# Add the project root directory to the Python path
sys.path.append(str(Path(__file__).resolve().parents[1]))

from btb_ai import app
from btb_ai.views.ai_model import image_to_data_uri


@pytest.fixture
def client():
    app.testing = True
    client = app.test_client()
    return client


def test_index(client):
    response = client.get("/")
    assert response.status_code == 200
    assert b"Hello" in response.data


def test_process_image(client):
    with requests_mock.Mocker() as mocker:
        # Mock the requests.post method and define the expected response
        expected_response = {"tags": "book, bookcase, bookshelf, fill, shelf, shelve"}
        mocker.post("http://localhost:5000/process_image", json=expected_response)

        # Send a POST request to the '/process-image' endpoint
        data = {"image_path": "imgs/photo_5440411542173633301_y.jpg"}
        response = client.post("/process_image", json=data)

        # Assert the response status code and content
        assert response.status_code == 200
        assert response.json == expected_response


def test_scene_explain(client):
    with requests_mock.Mocker() as mocker:
        # Mock the requests.post method and define the expected response
        with open("data/scene_explain_describe.json") as f:
            response = json.load(f)
        expected_response = {
            "classification": response["results"][0]["i18n"]["en"],
            "description": response["results"][0]["output"],
        }
        mocker.post("http://localhost:5000/scene_explain", json=expected_response)

        # Send a POST request to the '/scene_explain' endpoint
        img_path = "data/imgs/photo-1511108690759-009324a90311.jpeg"
        data = {
            "data": [
                {
                    "image": image_to_data_uri(img_path),
                    "features": ["question_answer"],
                    "style": "concise",
                    "question": "What is the thing in one word",
                },
            ]
        }
        response = client.post("/scene_explain", json=data)

        # Assert the response status code and content
        assert response.status_code == 200
        assert response.json == expected_response


btb_ai/views/ai_model.py
"""
Script contains AI models API that understand and describe the scene
"""
import base64
import json
import os
import requests

from dotenv import load_dotenv
from flask import jsonify, abort, request

import replicate
from jinaai import JinaAI

from btb_ai import app


# Load variables from .env file
load_dotenv()
# Access environment variables
scene_explain_token = os.environ.get("SCENE_EXPLAIN_TOKEN")


jinaai = JinaAI(
    secrets={
        "scenex-secret": scene_explain_token,
        "promptperfect-secret": "XXXXXX",
        "rationale-secret": "XXXXXX",
        "jinachat-secret": "XXXXXX",
    }
)


"""
@app.before_request
def send_image_to_process():
    "Imitation of a middleware"
    if request.path == '/process_image':
        # Assuming the image file is located on the server
        data = {'image_path':'data/ imgs/photo_5440411542173633301_y.jpg'}
        response = requests.post('http://localhost:5000/process_image', data=data) """


def image_to_data_uri(file_path):
    with open(file_path, "rb") as image_file:
        encoded_image = base64.b64encode(image_file.read()).decode("utf-8")
        return f"data:image/jpeg;base64,{encoded_image}"


def toBase64(img: str) -> str:
    return jinaai.utils.image_to_base64(f"{img}")


@app.route("/process_image", methods=["POST", "GET"])
def process_image():
    """
    Recognise Anything with Segment Anything using Replicate API
    https://github.com/xinyu1205/recognize-anything/tree/main
    https://github.com/IDEA-Research/Grounded-Segment-Anything
    """
    # TODO: img_path as request data
    # uncomment use the API, limited credits so using the dummy_response.json for now.
    """     image_path = 'data/imgs/photo_5440411542173633301_y.jpg'
    output = replicate.run(
            "idea-research/ram-grounded-sam:80a2aede4cf8e3c9f26e96c308d45b23c350dd36f1c381de790715007f1ac0ad",
            input={"input_image": open(image_path, "rb")}
        ) """
    with open("data/GSAM_response.json") as f:
        output = json.load(f)
    return jsonify({"tags": output["tags"]})


@app.route("/scene_explain", methods=["GET", "POST"])
def scene_explain():
    """
    https://scenex.jina.ai/
    https://github.com/jina-ai/jinaai-py/
    """
    # uncomment to use API
    #     img_path = 'data/imgs/photo-1511108690759-009324a90311.jpeg'
    #     data = {
    #     "data": [
    #         {"image": image_to_data_uri(img_path), "features": ["question_answer"], "style": "concise", "question": "What is the thing in one word"},
    #     ]
    # }
    #     response = jinaai.describe(data)
    with open("data/scene_explain_describe.json") as f:
        response = json.load(f)
    return jsonify(
        {
            "classification": response["results"][0]["i18n"]["en"],
            "description": response["results"][0]["output"],
        }
    )


@app.route("/")
def index():
    return "Hello"
    # abort(404)
